pkg/block/fetcher.go

```diff
@@ -392,33 +439,17 @@ func (f *BaseFetcher) fetchMetadata(ctx context.Context) (interface{}, error) {
 		})
 	}
 
-	partialBlocks := make(map[ulid.ULID]bool)
+	metaIDs, partialBlocks, err := f.blockIDsFetcher.GetActiveAndPartialBlockIDs(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "BlockIDsFetcher failed")
+	}
 	// Workers scheduled, distribute blocks.
 	eg.Go(func() error {
 		defer close(ch)
-		return f.bkt.Iter(ctx, "", func(name string) error {
-			parts := strings.Split(name, "/")
-			dir, file := parts[0], parts[len(parts)-1]
-			id, ok := IsBlockDir(dir)
-			if !ok {
-				return nil
-			}
-			if _, ok := partialBlocks[id]; !ok {
-				partialBlocks[id] = true
-			}
-			if !IsBlockMetaFile(file) {
-				return nil
-			}
-			partialBlocks[id] = false
-
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case ch <- id:
-			}
-
-			return nil
-		}, objstore.WithRecursiveIter)
+		for _, id := range metaIDs {
+			ch <- id
+		}
```

Comment:
The behavior seems to get changed. The previous implementation is streaming, but now it needs to wait for the response of `f.blockIDsFetcher.GetActiveAndPartialBlockIDs(ctx)` to start loading metadata.

Reply:
should we add the channel as a parameter to the interface function then?